lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/JsonRpcProvider.kt

```diff
@@ -44,65 +58,81 @@ class JsonRpcProvider(
         }
     }
 
-    private fun buildRequestJson(method: String, paramsJson: JsonElement): Map<String, JsonElement> {
-        val map = mapOf(
-            "jsonrpc" to JsonPrimitive("2.0"),
-            "method" to JsonPrimitive(method),
-            "id" to JsonPrimitive(0), // It is not used anywhere
-            "params" to paramsJson,
-        )
-
-        return JsonObject(map)
-    }
-
     private fun <T> buildRequest(
         method: JsonRpcMethod,
         paramsJson: JsonElement,
         responseSerializer: KSerializer<T>,
     ): HttpRequest<T> {
-        val requestJson = buildRequestJson(method.methodName, paramsJson)
+        val jsonRpcRequest = JsonRpcRequest(
+            id = 0,
+            jsonRpc = "2.0",
+            method = method.methodName,
+            params = paramsJson,
+        )
 
-        val payload =
-            HttpService.Payload(url, "POST", emptyList(), requestJson.toString())
+        return HttpRequest(url, jsonRpcRequest, responseSerializer, deserializationJson, httpService)
+    }
+
+    fun <T> batchRequests(requests: List<HttpRequest<T>>): BatchHttpRequest<T> {
+        require(requests.isNotEmpty()) { "Please provide requests while creating a batching requests" }
+
+        return buildBatchRequest(requests)
+    }
 
-        return HttpRequest(payload, buildJsonHttpDeserializer(responseSerializer, deserializationJson), httpService)
+    fun <T> batchRequests(vararg requests: HttpRequest<T>): BatchHttpRequest<T> {
+        require(requests.isNotEmpty()) { "Please provide requests while creating a batching requests" }
+
+        return buildBatchRequest(requests.toList())
+    }
```

Comment:
Can avoid some minor duplication here:
```suggestion
    fun <T> batchRequests(vararg requests: HttpRequest<T>): BatchHttpRequest<T> {
        return batchRequests(requests.toList())
    }
```
or
```suggestion
    fun <T> batchRequests(vararg requests: HttpRequest<T>): BatchHttpRequest<T> = batchRequests(requests.toList())
```

Reply:
Let's choose first suggestion 👍